internals/idkey/idkey.go

```diff
@@ -0,0 +1,265 @@
+func (k *IDKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	signer := k.key.(crypto.Signer)
+	return signer.Sign(rand, digest, opts)
+}
+
+// newEd25519 generates a new ed25519 key.
+func (k *IDKey) newEd25519() error {
+	var err error
+	_, k.key, err = ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// load loads the private identity key from storage.
+func (k *IDKey) load() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil || !exists {
+		return fmt.Errorf("directory %q is not accessible", k.keyDir)
+	}
+	err = expectPermission(k.keyDir, 0o700)
+	if err != nil {
+		return err
+	}
+	pemPath := filepath.Join(k.keyDir, identityKeyFile)
+	err = expectPermission(pemPath, 0o600)
+	if err != nil {
+		return err
+	}
+	// Load the key.
+	pemData, err := os.ReadFile(pemPath)
+	if err != nil {
+		return err
+	}
+	for {
+		var block *pem.Block
+		block, pemData = pem.Decode(pemData)
+		if block == nil {
+			break
+		}
+		switch block.Type {
+		case "PRIVATE KEY":
+			k.key, err = x509.ParsePKCS8PrivateKey(block.Bytes)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown PEM block %q in file %q", block.Type, pemPath)
+		}
+	}
+	if k.key == nil {
+		return fmt.Errorf("empty PEM file %q", pemPath)
+	}
+	return nil
+}
+
+// save saves the private identity key to storage.
+func (k *IDKey) save() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil {
+		return fmt.Errorf("directory %q is not accessible", k.keyDir)
+	}
+	if exists {
+		err = expectPermission(k.keyDir, 0o700)
+		if err != nil {
+			return err
+		}
+	} else {
+		// Create the leaf directory node with 0700 permissions.
+		err = os.Mkdir(k.keyDir, 0o700)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create the new private identity file.
+	pemPath := filepath.Join(k.keyDir, identityKeyFile)
+	pemFile, err := os.OpenFile(pemPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errors.Join(err, pemFile.Sync())
+		err = errors.Join(err, pemFile.Close())
+	}()
+	keyBytes, err := x509.MarshalPKCS8PrivateKey(k.key)
+	if err != nil {
+		return err
+	}
+	pemPrivateBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: keyBytes,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemPrivateBlock); err != nil {
+		return err
+	}
+	if _, err = pemFile.Write(pemBuffer.Bytes()); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Fingerprint returns the identity fingerprint. This is the SHA512/384 hash
+// of the public key, encoded in base32 (without padding). This is a
+// convenient shorthand form of the identity that can be used to identify
+// a specfic machine or device.
+func (k *IDKey) Fingerprint() string {
+	publicBytes := k.Public().(ed25519.PublicKey)
+	hashBytes := sha512.Sum384(publicBytes)
+	return base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(hashBytes[:])
+}
```

Comment:
Currently, most fingerprints are using `sha256` presented as an hexadecimal string. As an example, this is what Chrome will show when inspecting a certificate.

I would recommend at least using an hexadecimal string instead of base32 as it is more standard.

Reply:
Discussed this with @ghislainbourgeois.

I think this is a very interesting point.

Throughout the design process with the CTO, for the fingerprint, the focus was on CLI and the look, feel and consistency of the char set for the CLI user.

At attempt at using ed52219 keys on Chrome and Firefox actually highlights this is not yet supported (although I do know X25519 key exchange is supported).

We have a little time to refine this in the next cycle, so I propose we move forward, but I will make a not for discussion.

I see mTLS is also technically supported in browsers, so perhaps we should consider the use-case where a browser / phone / device actually connects to the API.